app/models/data_repo_client.rb

```diff
@@ -0,0 +1,207 @@
+##
+# API client bindings for retrieving information about Terra Data Repo datasets/snapshots/schemas from their API
+##
+class DataRepoClient < Struct.new(:access_token, :api_root, :storage, :expires_at, :service_account_credentials)
+  extend AccessTokenGenerator # class method for generating OAuth2 access tokens
+  include GoogleServiceClient # helper module for refreshing access tokens, accessing GCS storage driver attributes
+  include ApiHelpers # process external API requests/responses
+
+  # path to read-only service account JSON keyfile
+  SERVICE_ACCOUNT_KEY = !ENV['READ_ONLY_SERVICE_ACCOUNT_KEY'].blank? ? (ENV['NOT_DOCKERIZED'] ? ENV['READ_ONLY_SERVICE_ACCOUNT_KEY'] : File.absolute_path(ENV['READ_ONLY_SERVICE_ACCOUNT_KEY'])) : ''
+  # Google authentication scopes necessary for querying TDR API
+  GOOGLE_SCOPES = %w(openid email profile)
+  # GCP Compute project
+  COMPUTE_PROJECT = ENV['GOOGLE_CLOUD_PROJECT'].blank? ? '' : ENV['GOOGLE_CLOUD_PROJECT']
+  # Base API URL to request against
+  BASE_URL = Rails.application.config.tdr_api_base_url
+  # constant used for retry loops in process_firecloud_request and execute_gcloud_method
+  MAX_RETRY_COUNT = 5
+  # constant used for incremental backoffs on retries (in seconds); ignored when running unit/integration test suite
+  RETRY_INTERVAL = Rails.env.test? ? 0 : 15
+
+  # control variables
+  SORT_DIRECTIONS = %w(asc desc).freeze
+  SORT_OPTIONS = %w(name description created_date).freeze
+  DATASET_INCLUDE_FIELDS = %w(NONE SCHEMA ACCESS_INFORMATION PROFILE DATA_PROJECT STORAGE).freeze
+  ALL_DATASET_FIELDS = %w(SCHEMA ACCESS_INFORMATION PROFILE DATA_PROJECT STORAGE).freeze
+
+  ##
+  # Constructors & token management methods
+  ##
+
+  # initialize is called after instantiating with DataRepoClient.new
+  # will set the access token, TDR base api url root and GCP storage driver instance
+  #
+  # * *params*
+  #   - +service_account_key+: (String, Pathname) => Path to service account JSON keyfile
+  # * *return*
+  #   - +DataRepoClient+ object
+  def initialize(service_account=SERVICE_ACCOUNT_KEY)
+    # GCS storage driver attributes
+    storage_attr = {
+      project: COMPUTE_PROJECT,
+      timeout: 3600,
+      keyfile: service_account
+    }
+
+    self.service_account_credentials = service_account
+    self.access_token = self.class.generate_access_token(service_account)
+    self.storage = Google::Cloud::Storage.new(storage_attr)
+    self.expires_at = Time.zone.now + self.access_token['expires_in']
+    self.api_root = BASE_URL
+  end
+
+  ##
+  # Abstract request handlers
+  ##
+
+  # submit a request to TDR API
+  def process_api_request(http_method, path, payload: nil, retry_count: 0)
```

Comment:
this is a semantic nitpick, but `execute_api_request` reads better to me.  'process' sounds like we're just processing an already-sent request